deepchem/models/tensorgraph/layers.py

```diff
@@ -567,6 +589,32 @@ def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
     return out_tensor
 
 
+class Conditional(Layer):
+  """
+  A Conditional Layer wraps tf.Cond. 
+
+  Parameters
+  ----------
+  true_fn: function
+    Function to run if predicate evaluates to true.
+
+  false_fn: function
+    Function to run if predicate evaluates to false.
+
+  """
+  def __init__(self, true_fn, false_fn, **kwargs):
```

Comment:
I wonder if this is pickle-able (by wonder I mean doubt)

Can you leave a big TODO (yzhao)?

Reply:
Would it make more sense for this to have three input layers: (predicate, true value, false value)?  That way all the inputs would be specified by layers, not Tensorflow objects.

Also, it needs to compute `_shape`.

Reply:
The problem here is again, the nuanced condition that the thing that invokes create_tensor() must be the tf.cond itself. Otherwise it will pin a hard-dependency and build it into the graph. This is related to the docstring i mentioned above. If I create a layer explicitly, during the topological sort step, its create_tensor() will be invoked outside of the tf.Cond.

Reply:
Alternatively we'd have to modifiy the topological sort code so that it doesn't invoke conditional layers, but that's a little hairy.

Reply:
Is it a problem if `create_tensor()` gets called on both?  The issue described in the documentation didn't really seem relevant to TensorGraph.  It's important in Tensorflow because evaluating an operation can have side effects (like changing the value of a variable), so it matters a lot what does and doesn't get evaluated lazily.  But TensorGraph layers just take inputs and produce outputs.  They don't have side effects.

Reply:
It's not so much that create_tensor gets called on both so much as the expressions inside the false_fn and true_fn are actually computed and evaluated independent of the predicate if they are created OUTSIDE the tf.Cond scope.

The computation time of one of these branches is _extremely_ slow. While it's true that TensorGraph layers just take inputs and produce outputs, the topological sort and graph construction step will create the expressions inside true_fn and false_fn OUTSIDE of the tf.Cond name scope, which will cause them to be evaluated regardless of the predicate. The way it's currently written, it's tf.Cond itself that creates them INSIDE the scope.